### ISSUE TITLE
New postgresql exception in 1.6.99post4 (release candidate)

### ISSUE THREAD

**Issue by StevenCTimm** [HEPCloud/decisionengine#455]
2021-07-20 04:19:07,002 - AWS_Calculations_with_source_proxy - TaskManager - ERR
OR - {"event": "Exception running source SourceProxy ", "exception": "Traceback 
(most recent call last):\n  File \"/usr/lib/python3.6/site-packages/decisionengi
ne/framework/dataspace/datasources/postgresql.py\", line 259, in update\n    sel
f._select(q, (taskmanager_id, generation_id, key))[0]\nIndexError: list index ou
t of range\n\nDuring handling of the above exception, another exception occurred
:\n\nTraceback (most recent call last):\n  File \"/usr/lib/python3.6/site-packag
es/decisionengine/framework/taskmanager/TaskManager.py\", line 377, in run_sourc
e\n    self.data_block_put(data, header, self.data_block_t0)\n  File \"/usr/lib/
python3.6/site-packages/decisionengine/framework/taskmanager/TaskManager.py\", l
ine 312, in data_block_put\n    data_block.put(key, product, header, metadata=me
tadata)\n  File \"/usr/lib/python3.6/site-packages/decisionengine/framework/data
space/datablock.py\", line 287, in put\n    self._setitem(key, value, header, me
tadata=metadata)\n  File \"/usr/lib/python3.6/site-packages/decisionengine/frame
work/dataspace/datablock.py\", line 348, in _setitem\n    self._update(key, stor
e_value, header, metadata=metadata)\n  File \"/usr/lib/python3.6/site-packages/d
ecisionengine/framework/dataspace/datablock.py\", line 321, in _update\n    key
 value, header, metadata)\n  File \"/usr/lib/python3.6/site-packages/decisioneng
ine/framework/dataspace/dataspace.py\", line 120, in update\n    value, header, 
metadata)\n  File \"/usr/lib/python3.6/site-packages/decisionengine/framework/da
taspace/datasources/postgresql.py\", line 263, in update\n    taskmanager_id, ge
neration_id, key, table))\nKeyError: 'taskmanager_id=175 or generation_id=44 or 
key=Job_Limits not found in header'", "level": "error", "logger": "AWS_Calculati
ons_with_source_proxy", "module": "TaskManager", "timestamp": "2021-07-20 04:19:
07"}

-------

This is quite reproducible.  It has happened several times in various channels
in 24 hours.  The first one is in the AWS_Calculations_with_source_proxy channel
but other channels have been affected as well.
'

The question is, is the code supposed to be stable anymore with a postgresql data
source or must we switch to the SQLAlchemy data source?  It is not a problem to switch for testing but may be an issue if this code ever moves to production.

**Comment by StevenCTimm** [comment]
This error appears to happen most often at the end of the channel when the datablock is being PUT back into the database.  In several cases I have noted that we got all the way through the channel and classads were even advertised to the collector but then things crashed.  The channels most often affected are the ones which read from source proxies or write to them.  I have seen the error in question happen in 4 different channels.

**Comment by StevenCTimm** [comment]
At the moment the test node is installed with a SQLAlchemy-style data source.. if another native postgresql data source is needed, let me know and I will configure one.

Steve Timm

**Comment by mambelli** [comment]
The error seems similar to the SQLAlchemy one. I suspect that it may be more in the business logic (which data we save and look for) than in the DB implementation.

Here is the formatted traceback for convenience:
```
Traceback (most recent call last):
 File /usr/lib/python3.6/site-packages/decisionengine/framework/dataspace/datasources/postgresql.py, line 259, in update
 self._select(q, (taskmanager_id, generation_id, key))[0]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
 File /usr/lib/python3.6/site-packages/decisionengine/framework/taskmanager/TaskManager.py, line 377, in run_source
 self.data_block_put(data, header, self.data_block_t0)
 File /usr/lib/python3.6/site-packages/decisionengine/framework/taskmanager/TaskManager.py, line 312, in data_block_put
 data_block.put(key, product, header, metadata=metadata)
 File /usr/lib/python3.6/site-packages/decisionengine/framework/dataspace/datablock.py, line 287, in put
 self._setitem(key, value, header, metadata=metadata)
 File /usr/lib/python3.6/site-packages/decisionengine/framework/dataspace/datablock.py, line 348, in _setitem
 self._update(key, store_value, header, metadata=metadata)
 File /usr/lib/python3.6/site-packages/decisionengine/framework/dataspace/datablock.py, line 321, in _update
 keyvalue, header, metadata)
 File /usr/lib/python3.6/site-packages/decisionengine/framework/dataspace/dataspace.py, line 120, in update
 value, header, metadata)
 File /usr/lib/python3.6/site-packages/decisionengine/framework/dataspace/datasources/postgresql.py, line 263, in update
 taskmanager_id, generation_id, key, table))
KeyError: 'taskmanager_id=175 or generation_id=44 or key=Job_Limits not found in header'
```

**Comment by StevenCTimm** [comment]
Marco created 1.6.99post5 which did not have the sqlalchemy commits in it but otherwise similar to 1.6.99post4.  We see a very similar exception in that version, but note that the key is not always the same, it can be a different data frame that is missing

line 263, in update\n    taskmanager_id, generation_id, key, table))\nKeyError: 'taskmanager_id=64 or generation_id=37 or key=Factory_Entries_AWS not found in header'

it is therefore likely that this same error was in 1.6.99.post3 as well, and/or introduced by
the patch that Kyle added to fix the datablock ordering in transforms which the 1.6.99post3 RPM suffered from.

**Comment by StevenCTimm** [comment]
We have verified that (1) the exception is in Kyle's rpm where he made the fix of a different bug that was making transforms fail due to dependency ordering (2) that the exception happens not only with source proxies but also with sources, and always happens on the PUT of the datablock.. i.e. the source runs, gets its data, and throws the exception trying to put.

It is not possible to roll directly back to 1.6.99 post3 (one commit before that rpm ) because that too was broken
for a different reason.

**Comment by mambelli** [comment]
I prepared an instrumented 1.6.2 (1.6.2.post1) to inspect the behavior in that version. All changes are new log messages or additional checks resulting in log messages. Nothing affecting the behavior. Specifically, I instrumented the datasource postgresql.py. I compared the output of the changed SELECT_LAST_GENERATION_ID_BY_NAME query, I added the checks for taskmanager_id, generation_id, key in the different tables, I printed a warning if some queries, especially updates, had no match. If something was triggered a message starting w/ "DB probe -" and a stack trace were added to the log file (`log.exception()`).
After running for few hours I found several messages. Some of these messages correspond to verifications added in 1.7 and raising an exception there (e.g. "exception in update verification" raise KeyError). All errors/inconsistencies are in source proxies, the majority are SELECT_LAST_GENERATION_ID_BY_NAME_AND_ID not finding any record. There is no error from SELECT_LAST_GENERATION_ID_BY_NAME. The other error messages are the consistency check failing and updates matching 0 records (not updating any record).  Sometimes the triplet taskmanager_id, generation_id, key is in none of the tables (header, metadata, dataproduct), sometimes is missing only in header or the first 2, and the update succeeds.

In any case the DB inconsistencies are also in 1.6. 1.6.99.post4 is failing only because there are more checks. The question to clarify is if it is OK for these to be there (are to be expected with source proxies) or not.
